contracts/icp/context-config/src/types.rs

```diff
@@ -0,0 +1,360 @@
+            source: value.source.0.into_owned(),
+            metadata: value.metadata.0.into_inner().into_owned(),
+        }
+    }
+}
+
+impl<'a> From<ICApplication> for Application<'a> {
+    fn from(value: ICApplication) -> Self {
+        Application::new(
+            value.id.rt().expect("infallible conversion"),
+            value.blob.rt().expect("infallible conversion"),
+            value.size,
+            ApplicationSource(Cow::Owned(value.source)),
+            ApplicationMetadata(Repr::new(Cow::Owned(value.metadata))),
+        )
+    }
+}
+
+#[derive(CandidType, Serialize, Deserialize, Debug, Clone)]
+pub struct ContextRequest {
+    pub context_id: ICContextId,
+    pub kind: ContextRequestKind,
+}
+
+#[derive(CandidType, Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+pub enum ICCapability {
+    ManageApplication,
+    ManageMembers,
+    Proxy,
+}
+
+#[derive(CandidType, Serialize, Deserialize, Debug, Clone)]
+pub enum ContextRequestKind {
+    Add {
+        author_id: ICContextIdentity,
+        application: ICApplication,
+    },
+    UpdateApplication {
+        application: ICApplication,
+    },
+    AddMembers {
+        members: Vec<ICContextIdentity>,
+    },
+    RemoveMembers {
+        members: Vec<ICContextIdentity>,
+    },
+    Grant {
+        capabilities: Vec<(ICContextIdentity, ICCapability)>,
+    },
+    Revoke {
+        capabilities: Vec<(ICContextIdentity, ICCapability)>,
+    },
+    UpdateProxyContract,
+}
+
+#[derive(CandidType, Serialize, Deserialize, Debug, Clone)]
+pub enum RequestKind {
+    Context(ContextRequest),
+}
+
+#[derive(CandidType, Serialize, Deserialize, Debug, Clone)]
+pub struct Request {
+    pub kind: RequestKind,
+    pub signer_id: ICSignerId,
+    pub timestamp_ms: u64,
+}
+
+impl Request {
+    pub fn new(signer_id: ICSignerId, kind: RequestKind) -> Self {
+        Self {
+            signer_id,
+            kind,
+            timestamp_ms: 0, // Default timestamp for tests
+        }
+    }
+}
+
+#[derive(Debug, ThisError)]
+pub enum ICPSignedError<E> {
+    #[error("invalid signature")]
+    InvalidSignature,
+    #[error("json error: {0}")]
+    ParseError(#[from] serde_json::Error),
+    #[error("derivation error: {0}")]
+    DerivationError(E),
+    #[error("invalid public key")]
+    InvalidPublicKey,
+    #[error("signature error: {0}")]
+    SignatureError(#[from] ed25519_dalek::ed25519::Error),
+}
+
+#[derive(Deserialize, Debug, Clone)]
+struct Phantom<T>(#[serde(skip)] std::marker::PhantomData<T>);
+
+impl<T> CandidType for Phantom<T> {
+    fn _ty() -> candid::types::Type {
+        candid::types::TypeInner::Null.into()
+    }
+
+    fn idl_serialize<S>(&self, serializer: S) -> Result<(), S::Error>
+    where
+        S: candid::types::Serializer,
+    {
+        serializer.serialize_null(())
+    }
+}
+
+#[derive(CandidType, Deserialize, Debug, Clone)]
+pub struct ICPSigned<T> {
+    pub payload: Vec<u8>,
+    pub signature: Vec<u8>,
```

Comment:
```suggestion
    payload: Vec<u8>,
    signature: [u8; 32],
```

Reply:
I tried changing signature to [u8; 32] but ed25519 signatures always have 64 bytes so it throwed error when trying to parse it... 
When I switched to [u8; 64] type it gives error because of deseralization...

Reply:
> I tried changing signature to [u8; 32] but ed25519 signatures 

sorry, I meant `[u8; 64]`

> When I switched to [u8; 64] type it gives error because of deseralization

what was the error?